### PR TITLE
Feat: Added Security to Setting and App Passocde Splash Screen

### LINKED ISSUE #873: (Front) Settings / Security Section
**Description:**
The goal is to strengthen Mapeo’s security to guard against physical device seizure and surveillance. This screen is to add a Security button linking to the passcode screens.

**Screen:**
Security
App Passcode and Device Security (Copy TBD)

**Design:**
https://www.figma.com/file/ZOsM89L0xy0NUOCkuksJgq/OTF-%2B-UXFund?node-id=47960%3A908

Points : 2

### LINKED ISSUE #874: (Front) Settings / Security Screen
**Description:**
User is on the Settings screen, clicks on Security. User is taken to the security screen.
Screen will reflect whether a passcode has been set or not

**Screens (Copy TBD)**
### #1 No passcode set

Device Security

App Passcode
_Passcode not set_

Kill Passcode
To use, enable App Passcode

--
Device Backup

Paper Key
Reinstate your account if this device is lost

Design:
https://www.figma.com/file/ZOsM89L0xy0NUOCkuksJgq/OTF-%2B-UXFund?node-id=47960%3A916

### #2 Passcode set

Device Security

App Passcode
_Passcode is set_

Kill Passcode
To use, enable App Passcode

--
Device Backup

Paper Key
Reinstate your account if this device is lost

Design: https://www.figma.com/file/ZOsM89L0xy0NUOCkuksJgq/OTF-%2B-UXFund?node-id=47972%3A1197

Points : 3

### LINKED ISSUE #875: (Front) App Passcode / Landing Screen
**Description:**
User is on the Settings/Security screen and clicks on App Passcode, user is taken to the landing page
The App Passcode Landing page will always be displayed, whether the user has set a passcode or not

**Screen (Copy TBD):**
(Arrow to click back to Settings/Security screen)
What is App Passcode 
LOREM IPSUM

(Continue)

**Copydeck**
https://docs.google.com/spreadsheets/d/1od1MdXpS32VjTSGgQQmILp72YPXlvxbfuh95q_JxWTI/edit#gid=1176152038

**Design:**
https://www.figma.com/file/ZOsM89L0xy0NUOCkuksJgq/OTF-%2B-UXFund?node-id=47960%3A949

Points : 2

### PR DESCRIPTION
**Pull request by ErikSin**
## Description
* Added `Security` to settings screen
* Clicking `Security` Navigates to new Security Screen
* Clicking `App Passcode` in Security Screen Navigates to App Passcode Splash Page

Kapture 2021-12-04 at 14 38 25

## Wireframe
https://www.figma.com/file/ZOsM89L0xy0NUOCkuksJgq/OTF-%2B-UXFund?node-id=47898%3A770

## To Do
* Purely UI, no backend functionality yet
* App Passcode Creation Screens
* Kill Passcode Screens

### CODE REVIEW

File: report.20211204.125807.1973.0.001.json [hunk 1]
```diff
@@ -0,0 +1,961 @@
+{
```
Comment (achou11) [inline report.20211204.125807.1973.0.001.json]: this file shouldn't be included in the PR

File: src/frontend/context/AppProvider.js [hunk 2]
```diff
@@ -14,7 +15,9 @@ const AppProvider = ({ children }: { children: React.Node }) => (
     <ObservationsProvider>
       <ConfigProvider>
         <SettingsProvider>
-          <DraftObservationProvider>{children}</DraftObservationProvider>
+          <SecurityProvider>
```
Comment (achou11) [inline src/frontend/context/AppProvider.js:18]: we should really invest some time into a proper global state management solution at some point 😅 one day...

Comment (ErikSin) [reply src/frontend/context/AppProvider.js:18]: YES. I chuckled a bit when I read this comment. Our state management is abysmal....

File: src/frontend/screens/Security/index.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+          <ListItemText
+            primary={<FormattedMessage {...m.killPasscodeHeader} />}
+            secondary={<FormattedMessage {...killPassCodeDes} />}
+          />
+        </ListItem>
+
+        <ListItem style={[styles.divder]}>
+          {/* Divider, added a React Fragment as List Item Requires a child */}
+          <React.Fragment></React.Fragment>
+        </ListItem>
```
Comment (achou11) [inline src/frontend/screens/Security/index.tsx]: i'd look into adding a prop on the `ListItem` to enable a bottom/top border instead of creating a completely separate list item for a divider

File: src/frontend/screens/AppPasscode/NewPasscode/index.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import * as React from "react";
+import { View } from "react-native";
+import { SplashScreen } from "./Splash";
+
+enum ScreenState {
+  splash,
```
Comment (achou11) [inline src/frontend/screens/AppPasscode/NewPasscode/index.tsx:6]: nit but maybe rename to something closer to `intro`? i think of the app/os splashscreen when i see splash

File: src/frontend/screens/AppPasscode/NewPasscode/index.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    return <SplashScreen incrementState={incrementState} />;
+  }
+
+  //To-Do Create Set Passcode Screen
+  if (screenState === ScreenState.setPasscode) {
+    return <React.Fragment></React.Fragment>;
+  }
+
+  //To-Do Create Confirm passcode Screen
+  return <React.Fragment></React.Fragment>;
```
Comment (achou11) [inline src/frontend/screens/AppPasscode/NewPasscode/index.tsx:34]: Are these not going to be built in this PR?

File: src/frontend/screens/AppPasscode/NewPasscode/index.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  setPasscode,
+  reenterPasscode,
+}
+
+export const NewPasscode = () => {
+  const [screenState, setScreenState] = React.useState<ScreenState>(
+    ScreenState.splash
+  );
+
+  function incrementState() {
```
Comment (achou11) [inline src/frontend/screens/AppPasscode/NewPasscode/index.tsx]: nit but maybe rename to `nextStep`? Not sure if addressed in designs but would be good to handle going to previous steps too right?

File: src/frontend/screens/AppPasscode/index.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  },
+});
+
+export const AppPasscode: NavigationStackScreenComponent = () => {
+  const { passIsSet } = React.useContext(SecurityContext);
+
+  return (
+    <View style={styles.pageContainer}>
+      {!passIsSet && <NewPasscode />}
+      {passIsSet && <EnterPasscode />}
```
Comment (achou11) [inline src/frontend/screens/AppPasscode/index.tsx]: use a ternary here?

File: src/frontend/screens/Security/SecurityContext.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  setPassIsSet: React.Dispatch<React.SetStateAction<boolean>>;
+}
+
+const defaultSecurityContext = [false, () => {}];
+
+//@ts-ignore
+export const SecurityContext = React.createContext<SecurityContextType>(
+  //@ts-ignore
+  defaultSecurityContext
+);
```
Comment (achou11) [inline src/frontend/screens/Security/SecurityContext.tsx:15]: what's the reason for the ignores here?

**Comment by achou11** [comment]
Overall looks good with a a few comments. As we discussed offline, probably makes sense to put this behind a feature flag, so will re-review when that happens

File: src/frontend/screens/AppPasscode/NewPasscode/index.tsx (hunk 5 above)
Comment (ErikSin) [reply src/frontend/screens/AppPasscode/NewPasscode/index.tsx:34]: Not in this PR

File: src/frontend/screens/AppPasscode/NewPasscode/index.tsx (hunk 6 above)
Comment (ErikSin) [reply src/frontend/screens/AppPasscode/NewPasscode/index.tsx]: I added the ability to change to a previous step, but I left it as incrementState() because I couldn't think of a word that captures next and back

File: src/frontend/screens/Security/SecurityContext.tsx (hunk 8 above)
Comment (ErikSin) [reply src/frontend/screens/Security/SecurityContext.tsx:15]: My `SecurityContextType` Typing is very specific, that the default security context type cannot be captured in it. BUT since my context is initialzed here as well, the default isnt ever going to be used. Also this pattern is most likely going to change once we implement it so I think ts-ignoring is the most graceful way of dealing with it right now

File: src/frontend/screens/AppPasscode/index.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+});
+
+export const AppPasscode: NavigationStackScreenComponent = () => {
+  const { passIsSet } = React.useContext(SecurityContext);
+  const [{ experiments }] = React.useContext(SettingsContext);
+  const { navigate } = useNavigation();
+
+  React.useEffect(() => {
+    if (!experiments.appPasscode) navigate("Settings");
+  }, []);
```
Comment (achou11) [inline src/frontend/screens/AppPasscode/index.tsx]: `experiments` should be added to the deps array here

Comment (achou11) [reply src/frontend/screens/AppPasscode/index.tsx]: technically `navigate` should too but i know that there are some issues with adding that in a useEffect (i.e. causing a ton of re-runs), so okay to omit that one here

File: src/frontend/screens/Security/index.tsx [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+});
+
+export const Security: NavigationStackScreenComponent = () => {
+  const { passIsSet } = React.useContext(SecurityContext);
+  const [{ experiments }] = React.useContext(SettingsContext);
+  const { navigate } = useNavigation();
+
+  React.useEffect(() => {
+    if (!experiments.appPasscode) navigate("Settings");
+  }, []);
```
Comment (achou11) [inline src/frontend/screens/Security/index.tsx]: similar comment about updating the deps array to include `experiments`

File: src/frontend/screens/Security/index.tsx [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+  const [passCodeDes, killPassCodeDes] = React.useMemo(
+    () =>
+      passIsSet
+        ? [m.passDesriptionPassSet, m.killPassDescriptonPassSet]
+        : [m.passDesriptionPassNotSet, m.killPassDescriptonPassNotSet],
+    [passIsSet]
+  );
```
Comment (achou11) [inline src/frontend/screens/Security/index.tsx:81]: couple of nits:

- is the memoization necessary here? guessing it may not be but if there's a noticeable behavior it prevents, fine to keep
- probably would make sense to extract the arrays as constants defined outside of the component e.g.

```ts
// Outside of the Security component
const PW_DESCS_WHEN_SET = [m.passDesriptionPassSet, m.killPassDescriptonPassSet]
const PW_DESCS_WHEN_NOT_SET = [m.passDesriptionPassNotSet, m.killPassDescriptonPassNotSet]

export const Security = () => {
  // ... 
  const [passCodeDes, killPassCodeDes] = React.useMemo(
    () =>
      passIsSet
        ? PASSWORD_DESCRIPTION_SET
        : PASSWORD_DESCRIPTION_NOT_SET
    [passIsSet]
  );
}
```

File: src/frontend/sharedComponents/List/ListItem.tsx [hunk 12]
```diff
@@ -2,9 +2,22 @@ import React from "react";
 import { StyleSheet, View } from "react-native";
 import PropTypes from "prop-types";
 
-import { TouchableNativeFeedback } from "../../sharedComponents/Touchables";
+import { TouchableNativeFeedback } from "../Touchables";
 import { VERY_LIGHT_BLUE } from "../../lib/styles";
 import ListContext from "./ListContext";
+import { ViewStyleProp } from "../../sharedTypes";
+
+interface ListItemProp {
```
Comment (achou11) [inline src/frontend/sharedComponents/List/ListItem.tsx:10]: nit: rename to `ListItemProps`

File: src/frontend/sharedComponents/List/ListItem.tsx [hunk 13]
```diff
@@ -117,4 +131,30 @@ ListItem.propTypes = {
   divider: PropTypes.bool,
 };
 
-export default ListItem;
+interface ListDividerProp {
```
Comment (achou11) [inline src/frontend/sharedComponents/List/ListItem.tsx:134]: nit: rename to `ListDividerProps`

File: src/frontend/sharedComponents/List/ListItemText.tsx [hunk 14]
```diff
@@ -2,6 +2,15 @@ import React from "react";
 import { StyleSheet, View } from "react-native";
 import Text from "../Text";
 import PropTypes from "prop-types";
+import { TextStyleProp } from "../../sharedTypes";
+import { FormattedMessage } from "react-intl";
+
+interface ListItemTextProp {
```
Comment (achou11) [inline src/frontend/sharedComponents/List/ListItemText.tsx:8]: nit: rename to `ListItemTextProps`

**Review by achou11 — APPROVED** [review]
left a few comments that require some small changes, but overall lgtm 👍
